src/Search/Extensions/IndexableExtension.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace SilverStripe\FullTextSearch\Search\Extensions;
+
+use SilverStripe\ORM\DataExtension;
+
+/**
+ * Checks if a DataObject is publically viewable thus able to be added or retrieved from a publically searchable index
+ */
+class IndexableExtension extends DataExtension
```

Comment:
Making this a default extension on `DataObject` means more overhead for every method call, and framework boot times. It also makes this hard to extend by modules and custom implementations. For example, how would subsites indicate that a record shouldn't be indexed if a subsite has been "disabled" (hypothetically). I'd make this a service instead, and by extending ViewableData you can allow extension hooks on it. By checking for an `isIndexable` method on the DataObject itself, you can also allow more fine tuning of this functionality in a way that's hard through an extension. You're only using public methods on DataObjects here anyway. By making it a singleton, you can add cache state (`$indexableState`) in the service rather than on the DataObject

Reply:
@chillu 

service/singleton makes sense, will do that

> by extending ViewableData 
hmm, a bit strange.  reads weird and will end up with a bunch of unnecessary function.  wouldn't it be cleaner to just add the Extensible trait directly (along with the Injectable trait)?

So regarding making this extendable, is this how you see it working?

```
class IndexableService
..
use Extensible;
... 
public function isIndexable(DataObject $obj)
{
    $result = [do stuff]
    $this->extend('updateIsIndexable', $obj, $result);
}
```

Add in say subsites module:

extensions.yml
```
---
Name: subsitesfulltextsearch
Only:
  moduleexists: 'silverstripe/fulltextsearch'
---
SilverStripe\FullTextSearch\Search\Services\IndexableService:
  extensions:
    - SilverStripe\Subsites\Extensions\SubsitesIndexableServiceExtension
```

SubsitesIndexableServiceExtension.php
```
class SubsitesIndexableServiceExtension extends Extension
...
public function updateIsIndexable(&obj, &result)
{
    if ($subsiteDisabled) {
        $result = false;
```

Let me know if my thinking is correct or needs adjusting

Reply:
That all makes sense to me. Agree that using `Extensible` / `Injectable` / `Configurable` traits should be sufficient.

Reply:
Have refactored this from a DataExtension to a Singleton/Service

Reply:
Sorry I mean `Extensible`, this comes from an old school time when this was all lumped into `ViewableData` back in 3.x :D `extend()` also by default allows for NULL returns, so you can have extensions "abstain" from a decision which can be important. Same logic as in canView etc